### PR TITLE
Improvements to DHCP client

### PR DESCRIPTION
**Pull request by troglobit**
This PR extends the functionality of the DHCPv4 client by adding support for:
 - Changing the default request options
 - Acting on more options: NTP servers, routes, DNS servers
 - Configurable ARP:ing of acquired lease, default: enabled
 - Configurable route preference (metric) per DHCP interface.  Applies to options 3, 121, 249

All routes, including default route, is by default set with proto `dhcp` and metric 100.  This ensure that routes from other routing protocols, as well as statically set routes, have priority over DHCP acquired routes.

DNS servers, and search/domain settings, are saved per interface and enabled globally in the system using the dnsmasq resolver functionality.  A statically set DNS server will always take precedence.

NTP servers are added as non-preferred sources, and will be selected according to the NTPv4 protocol (strata), unless there is a statically configured NTP server with the `prefer` flag set.

A set of basic DHCP tests have been added to verify that leases are accepted and routes are installed properly (according to option precedence rules defined in RFC).  There are currently no regression tests for DNS or NTP yet, these can easily be added later when we have operational support for querying system DNS and NTP servers.  (Manual testing has of course been performed.)

> **Note:** for NTP servers to actually *take*, the ietf-system NTP client must be enabled.  (It can be enabled without any static servers set, but is by default disabled.)

### CODE REVIEW

**Comment by troglobit** [comment]
(Re: force push: just squashed and moved around a few commits to reduce the amount of cruft.)

**Review by mattiaswal — COMMENTED** [review]
One documentation change and one question.

File: src/confd/src/infix-dhcp.c [hunk 1]
```diff
@@ -13,44 +13,276 @@
 #include <srx/srx_val.h>
 
 #include "core.h"
+#define  ARPING_MSEC  1000
+#define  MODULE       "infix-dhcp-client"
+#define  XPATH        "/infix-dhcp-client:dhcp-client"
 
-static const struct srx_module_requirement infix_dhcp_reqs[] = {
-	{ .dir = YANG_PATH_, .name = "infix-dhcp-client", .rev = "2023-05-22" },
+static const struct srx_module_requirement reqs[] = {
```
[the remark this answers is not in the document]
Comment (troglobit) [reply src/confd/src/infix-dhcp.c:20]: this is just the auto-loader, the dependency-order handling is done by the bootstrap script.  for models we don't have any dependencies (or need factory settings) we can load using the auto-loader (`srx_require_modules()`) in each .C file's init function.

File: board/netconf/rootfs/libexec/infix/cli-pretty [hunk 2]
```diff
@@ -375,7 +375,7 @@ def ietf_routing(json, ip="ipv4"):
         sys.exit(1)
     hdr = (f"{'PREFIX':<{PadRoute.prefix}}"
            f"{'NEXT-HOP':<{PadRoute.next_hop}}"
-           f"{'METRIC':<{PadRoute.metric}}"
+           f"{'PREF':>{PadRoute.pref}}  "
```
[the remark this answers is not in the document]
Comment (troglobit) [reply board/netconf/rootfs/libexec/infix/cli-pretty:378]: Ah, good point. Totally forgot about that, thanks!

[the remark this answers is not in the document]
Comment (troglobit) [reply board/netconf/rootfs/libexec/infix/cli-pretty:378]: Fixed in latest force-push.

**Review by mattiaswal — APPROVED** [review]

**Pull request merged**